src/components/utils/src/timer.cc

```diff
@@ -135,37 +59,75 @@ timer::Timer::Timer(const std::string& name, const TimerTask* task_for_tracking)
 timer::Timer::~Timer() {
   LOG4CXX_AUTO_TRACE(logger_);
   LOG4CXX_DEBUG(logger_, "Timer is to be destroyed " << name_);
-  Stop();
   sync_primitives::AutoLock auto_lock(task_lock_);
   DCHECK(task_);
   delete task_;
+  if (thread_) {
+    thread_->join();
+    DCHECK(delegate_);
+    delete delegate_;
+    threads::DeleteThread(thread_);
+  }
 }
 
 void timer::Timer::Start(const Milliseconds timeout, const bool repeatable) {
   LOG4CXX_AUTO_TRACE(logger_);
   sync_primitives::AutoLock auto_lock(lock_);
-  SetTimeoutUnsafe(timeout);
-  repeatable_ = repeatable;
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
+
+  // There would be no way to stop thread if timeout in lopper will be 0
+  timeout_ms_ = (0u != timeout) ? timeout : 1u;
+
+  // It's new timer and we have to create it
+  if (!delegate_ && !thread_) {
+    delegate_ =
+        repeatable ? new TimerLooperDelegate(this) : new TimerDelegate(this);
+
+    DCHECK(delegate_);
+    thread_ = threads::CreateThread(name_.c_str(), delegate_);
+    DCHECK(thread_);
   }
-  StartUnsafe();
+  // else timer has been already starting
+
+  if (thread_->is_running()) {
+    LOG4CXX_INFO(logger_, "Restart timer in thread " << name_);
+    delegate_->ShouldBeRestarted();
+    UpdateTimeOut(timeout_ms_);
+  } else {
+    UpdateTimeOut(timeout_ms_);
+    thread_->start();
+  }
+
+  delegate_->WaitUntilStart();
 }
 
 void timer::Timer::Stop() {
   LOG4CXX_AUTO_TRACE(logger_);
   sync_primitives::AutoLock auto_lock(lock_);
-  repeatable_ = false;
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK(stop_result);
+  DCHECK(thread_);
+  LOG4CXX_DEBUG(logger_, "Stopping timer  " << name_);
+  if (pthread_equal(pthread_self(), thread_->thread_handle())) {
+    // Stop itself can't, so suspend
+    Suspend();
+  } else {
+    thread_->join();
   }
 }
 
 bool timer::Timer::IsRunning() const {
   sync_primitives::AutoLock auto_lock(lock_);
-  return is_running_;
+  if (!thread_)
+    return false;
+  DCHECK(delegate_);
```

Comment:
No. It `thread == NULL` it's ok, but if `thread != NULL && delegate_ == NULL` oh, it's huge problem and mistake in logic.